content/en/docs/concepts/scheduling-eviction/taint-and-toleration.md

```diff
@@ -85,9 +85,26 @@ An empty `effect` matches all effects with key `key1`.
 {{< /note >}}
 
 The above example used `effect` of `NoSchedule`. Alternatively, you can use `effect` of `PreferNoSchedule`.
-This is a "preference" or "soft" version of `NoSchedule` -- the system will *try* to avoid placing a
-pod that does not tolerate the taint on the node, but it is not required. The third kind of `effect` is
-`NoExecute`, described later.
+
+
+Allowed values for taint `effects` are:
+
+`NoExecute`
+: This affects pods that are already running on the node as follows
+    * pods that do not tolerate the taint are evicted immediately
+    * pods that tolerate the taint without specifying `tolerationSeconds` in
+   their toleration specification remain bound forever
+    * pods that tolerate the taint with a specified `tolerationSeconds` remain
+   bound for the specified amount of time
+   
+`NoSchedule`
+: No new pods will be scheduled on the node unless they have the matching toleration. Pods that are currently running on the node are not evicted.
+
+`PreferNoSchedule`
+: This is a "preference" or "soft" version of `NoSchedule`. The system will *try* to avoid placing a
+pod that does not tolerate the taint on the node, but it is not required. 
+
+All the taint effects are enforced by Kubernetes scheduler except for `NoExecute` which is enforced by NodeController.
```

Comment:
Is this right?
```suggestion
All the taint effects are enforced by the Kubernetes scheduler except for `NoExecute` which is
enforced by the kubelet.
```
It might actually be the kube-controller-manager (I haven't checked).

Reply:
I am not sure too, this is from the[ code ](https://github.com/kubernetes/kubernetes/blob/2c6c4566eff972d6c1320b5f8ad795f88c822d09/pkg/apis/core/types.go#L2992). Happy to remove it altogether

Reply:
Lets leave this paragraph out for now, that's the simple option.

Reply:
It was the node lifecycle controller within the kube-controller-manager